### PR TITLE
[Integration][Wiz] Task ti2coe/add parallelism for wiz vulnerabilities ingestion

### PR DESCRIPTION
**Pull request by VenfiOranai**
# Description

**What** — Add optional parallel pagination for Wiz vulnerability findings resync.

**Why** — Full resyncs on large tenants are slow with a single sequential cursor chain.

**How** — When `parallelism` is set on the vulnerability-finding selector, the client splits the query into `firstSeenAt` date windows or severity shards, probes each partition via `totalCount`, splits oversized partitions, and crawls them concurrently through `ReadyPartitionCrawlStream`. A token-bucket rate limiter caps Wiz API throughput (default 10 req/s). Behavior is unchanged when `parallelism` is omitted.

```yaml
parallelism:
  strategy: auto          # auto | date | severity
  lookbackDays: 365
  dateIntervalDays: 30
  apiRequestsPerSecond: 10
  maxPartitionEntities: 500
```

## Type of change

- [x] New feature (non-breaking change which adds functionality)

### Preflight checklist

- [x] Handled rate limiting
- [x] Handled pagination
- [x] Implemented the code in async

---

> [!NOTE]
> **Medium Risk**
> Raises concurrent Wiz API load and relies on partition filters for correctness/completeness; behavior is opt-in but changes pagination limits and ingestion paths for vulnerability findings.
> 
> **Overview**
> Adds **optional parallel pagination** for Wiz **vulnerability findings** resync so large result sets can be fetched with multiple concurrent GraphQL cursor chains instead of a single sequential paginator.
> 
> Integration config gains a **`parallelism`** block on the vulnerability-finding selector (`maxConcurrent`, `strategy` `auto`/`date`/`severity`, `dateIntervalDays`, `lookbackDays`). Resync passes that into client options; when set, **`WizClient`** splits the query into **partitions** (date windows on `firstSeenAt`/`updatedAt`, or severity shards) and runs **`_fetch_cursor_chain`** per partition under a semaphore, merging partition filters and dropping the shared cursor.
> 
> New **`wiz/pagination`** helpers implement partition building and filter merge; **`max_pages`** stopping logic is adjusted (`page_num > max_pages`). Unit tests cover partition fan-out, parallelism wiring, and strategy edge cases.
> 
> <sup>Reviewed by [Cursor Bugbot](https://cursor.com/bugbot) for commit 5d6e3785b60556b9c63327165c10fd8db1084e1f. Configure [here](https://www.cursor.com/dashboard/bugbot).</sup>

### CODE REVIEW

File: integrations/wiz/main.py [hunk 1]
```diff
@@ -93,16 +93,24 @@ async def resync_vulnerability_findings(kind: str) -> ASYNC_GENERATOR_RESYNC_TYP
...
+        "status_list": status_list,
+        "severity_list": severity_list,
+    }
+    if parallelism_selector is not None:
+        options["parallelism"] = {
+            "max_concurrent": parallelism_selector.max_concurrent,
+            "strategy": parallelism_selector.strategy,
+            "date_interval_days": parallelism_selector.date_interval_days,
+            "lookback_days": parallelism_selector.lookback_days,
+        }
```
[the remark this answers is not in the document]
Comment (VenfiOranai) [reply integrations/wiz/main.py]: Doing that right before merge to minimize conflicts

File: integrations/wiz/wiz/pagination/utils.py [hunk 2]
```diff
@@ -0,0 +1,92 @@
...
+    current_time = now or datetime.datetime.now(datetime.UTC)
+    start_time = current_time - datetime.timedelta(days=lookback_days)
+    interval = datetime.timedelta(days=interval_days)
+
+    windows: list[tuple[datetime.datetime, datetime.datetime]] = []
+    window_start = start_time
+    while window_start < current_time:
+        window_end = min(window_start + interval, current_time)
+        windows.append((window_start, window_end))
+        window_start = window_end
```
[the remark this answers is not in the document]
Comment (VenfiOranai) [reply integrations/wiz/wiz/pagination/utils.py:22]: Added min 1 validation

File: integrations/wiz/wiz/pagination/vulnerability_findings.py [hunk 3]
```diff
@@ -0,0 +1,45 @@
+from typing import Any
+
+from wiz.options import ParallelismConfig
+
+from wiz.pagination.base import PaginationPartition, PartitionStrategy
+from wiz.pagination.utils import build_date_partitions
+
+VULNERABILITY_FINDING_SEVERITIES = ("CRITICAL", "HIGH", "MEDIUM", "LOW", "NONE")
```
Comment (Hadar-Vaknin) [inline integrations/wiz/wiz/pagination/vulnerability_findings.py]: This const list is duplicated in severity_list selector, let's move to shared place and import it.

File: integrations/wiz/overrides.py [hunk 4]
```diff
@@ -97,6 +97,54 @@ class ProjectResourceConfig(ResourceConfig):
...
+        ge=1,
+    )
+    lookback_days: int = Field(
+        alias="lookbackDays",
+        title="Lookback Days",
+        description="How many days back to partition when using date strategy.",
+        default=365,
+        ge=1,
+    )
+    api_requests_per_second: int = Field(
```
Comment (Hadar-Vaknin) [inline integrations/wiz/overrides.py:126]: The whole parallelism selector is totally reasonable,  but is there any reason why the customer needs to be involved in setting the API req/sec?
I mean, IMO API rate limiting shouldn't be user-owned configuration.
Let me know what u think

Comment (VenfiOranai) [reply integrations/wiz/overrides.py:126]: In general I do agree, but I think it might be a bit safer to allow it due to how Wiz's tokens work.
They have this thing where you have a limit of 10 requests per second per token, but 100 requests per second for the account. If they have a bunch of other tokens for various other things, I wouldn't want to limit them without giving some sort of workaround.

File: integrations/wiz/wiz/pagination/vulnerability_findings.py (hunk 3 above)
Comment (VenfiOranai) [reply integrations/wiz/wiz/pagination/vulnerability_findings.py]: Do you mean the typing on the `severity_list` field? If so I could convert it to a StrEnum, as is it's a tuple so I can't really use it for typing.
That sound good to you?

Comment (Hadar-Vaknin) [reply integrations/wiz/wiz/pagination/vulnerability_findings.py]: Absolutely

File: integrations/wiz/main.py [hunk 5]
```diff
@@ -93,19 +94,37 @@ async def resync_vulnerability_findings(kind: str) -> ASYNC_GENERATOR_RESYNC_TYP
...
+        options.parallelism = ParallelismConfig(
+            strategy=parallelism_selector.strategy,
+            date_interval_days=parallelism_selector.date_interval_days,
+            lookback_days=parallelism_selector.lookback_days,
+            api_requests_per_second=parallelism_selector.api_requests_per_second,
+            max_partition_entities=parallelism_selector.max_partition_entities,
+        )
+
+    upsert_batch: list[dict[str, Any]] = []
+    upsert_batch_size = 100
```
Comment (Hadar-Vaknin) [inline integrations/wiz/main.py]: Let's make it constant

File: integrations/wiz/wiz/pagination/partitioning/refiner.py [hunk 6]
```diff
@@ -0,0 +1,89 @@
...
+from wiz.pagination.partitioning.splitter import PartitionSplitter
+
+if TYPE_CHECKING:
+    from wiz.client import WizClient
+
+
+class PartitionRefiner:
+    def __init__(self, client: WizClient, config: ParallelismConfig) -> None:
+        self._client = client
+        self._config = config
```
Comment (Hadar-Vaknin) [inline integrations/wiz/wiz/pagination/partitioning/refiner.py:22]: the cofig is not used

File: integrations/wiz/wiz/pagination/partitioning/refiner.py [hunk 7]
```diff
@@ -0,0 +1,89 @@
...
+        max_entities = config.max_partition_entities
+        if count <= max_entities:
+            logger.info(
+                f"Partition {partition.label} has {count} entities (within limit of {max_entities})"
+            )
+            yield partition
+            return
+
+        logger.info(
+            f"Partition {partition.label} has {count} entities, splitting (limit {max_entities})"
```
Comment (Hadar-Vaknin) [inline integrations/wiz/wiz/pagination/partitioning/refiner.py:65]: Every partition will log on info level, which may be a lot of logs for high-volume syncs.
Worth changing some to debug, but it is up to u.

File: integrations/wiz/wiz/pagination/partitioning/splitter.py [hunk 8]
```diff
@@ -0,0 +1,86 @@
...
+
+    def bisect_date_partition(
+        self, partition: PaginationPartition
+    ) -> list[PaginationPartition] | None:
+        date_filter = partition.filter_overlay.get("firstSeenAt")
+        if not isinstance(date_filter, dict):
+            return None
+
+        after_raw = date_filter.get("after")
+        before_raw = date_filter.get("before")
```
Comment (Hadar-Vaknin) [inline integrations/wiz/wiz/pagination/partitioning/splitter.py:20]: I think it is worth checking that after >= before just to be sure

Comment (VenfiOranai) [reply integrations/wiz/wiz/pagination/partitioning/splitter.py:20]: That's covered by the `window_end - window_start <= self.MIN_DATE_WINDOW` check, added an explicit test for that scenario though

File: integrations/wiz/wiz/pagination/partitioning/refiner.py (hunk 7 above)
Comment (VenfiOranai) [reply integrations/wiz/wiz/pagination/partitioning/refiner.py:65]: Changed

File: integrations/wiz/wiz/pagination/partitioning/refiner.py (hunk 6 above)
Comment (VenfiOranai) [reply integrations/wiz/wiz/pagination/partitioning/refiner.py:22]: Changed calls to use this rather than having config injected into the functions

File: integrations/wiz/main.py (hunk 5 above)
Comment (VenfiOranai) [reply integrations/wiz/main.py]: Done

File: integrations/wiz/wiz/pagination/vulnerability_findings.py (hunk 3 above)
Comment (VenfiOranai) [reply integrations/wiz/wiz/pagination/vulnerability_findings.py]: Epic, done

**Review by Hadar-Vaknin — APPROVED** [review]
